File: apps/nongyu-go-track-server/internal/presence/scanner.go
package presence

import (
	"context"
	"database/sql"
	"log/slog"
	"time"

	"nongyu-go-track-server/internal/store/sqlite"
	"nongyu-go-track-server/internal/usersync"
)

type Scanner struct {
	store   *sqlite.Store
	syncer  *usersync.Syncer
	offline time.Duration
	log     *slog.Logger
	now     func() time.Time
	stop    chan struct{}
	done    chan struct{}
}

func NewScanner(store *sqlite.Store, syncer *usersync.Syncer, offlineAfter time.Duration, log *slog.Logger) *Scanner {
	return &Scanner{
		store:   store,
		syncer:  syncer,
		offline: offlineAfter,
		log:     log,
		now:     time.Now,
		stop:    make(chan struct{}),
		done:    make(chan struct{}),
	}
}

func (s *Scanner) Start() {
	go s.loop()
}

func (s *Scanner) Stop() {
	close(s.stop)
	<-s.done
}

func (s *Scanner) loop() {
	defer close(s.done)
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()
	s.ScanOnce()
	for {
		select {
		case <-s.stop:
			return
		case <-ticker.C:
			s.ScanOnce()
		}
	}
}

// ScanOnce 将超时用户置离线并回写 Node。
func (s *Scanner) ScanOnce() {
	now := s.now()
	cutoff := now.UTC().UnixMilli() - s.offline.Milliseconds()
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	ids, err := s.store.ListTimedOut(ctx, cutoff)
	if err != nil {
		s.log.Error("presence scan failed", "err", err)
		return
	}
	if len(ids) == 0 {
		return
	}
	ms := now.UTC().UnixMilli()
	err = s.store.WithWriteTx(ctx, func(tx *sql.Tx) error {
		for _, id := range ids {
			if err := s.store.UpsertPresence(ctx, tx, sqlite.Presence{
				UserID:       id,
				Online:       false,
				LastSeenAtMs: cutoff,
				UpdatedAtMs:  ms,
			}); err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		s.log.Error("presence offline write failed", "err", err)
		return
	}
	for _, id := range ids {
		s.syncer.Notify(id, false, ms)
	}
}

func SetOffline(ctx context.Context, store *sqlite.Store, syncer *usersync.Syncer, userID int64, now time.Time) error {
	ms := now.UTC().UnixMilli()
	err := store.WithWriteTx(ctx, func(tx *sql.Tx) error {
		return store.UpsertPresence(ctx, tx, sqlite.Presence{
			UserID:       userID,
			Online:       false,
			LastSeenAtMs: ms,
			UpdatedAtMs:  ms,
		})
	})
	if err != nil {
		return err
	}
	syncer.Notify(userID, false, ms)
	return nil
}

File: apps/nongyu-node-track-server/src/presence/scanner.ts
import type { Logger } from "pino";
import type { Store } from "../store/sqlite/db.js";
import { listTimedOut, upsertPresence } from "../store/sqlite/presence.js";
import type { Syncer } from "../usersync/index.js";

export class Scanner {
  private stopped = false;
  private loopPromise: Promise<void> | null = null;
  now: () => Date = () => new Date();

  constructor(
    private store: Store,
    private syncer: Syncer,
    private offlineAfterMs: number,
    private log: Logger,
  ) {}

  start(): void {
    this.loopPromise = this.loop();
  }

  async stop(): Promise<void> {
    this.stopped = true;
    if (this.loopPromise) await this.loopPromise;
  }

  private async loop(): Promise<void> {
    this.scanOnce();
    while (!this.stopped) {
      await sleep(60_000);
      if (this.stopped) break;
      this.scanOnce();
    }
  }

  /** 将超时用户置离线并回写 Node。 */
  scanOnce(): void {
    const now = this.now();
    const cutoff = now.getTime() - this.offlineAfterMs;
    try {
      const ids = listTimedOut(this.store, cutoff);
      if (ids.length === 0) return;
      const ms = now.getTime();
      this.store.withWriteTx(() => {
        for (const id of ids) {
          upsertPresence(this.store, {
            userId: id,
            online: false,
            lastSeenAtMs: cutoff,
            platform: "",
            appVersion: "",
            deviceBrand: "",
            updatedAtMs: ms,
          });
        }
      });
      for (const id of ids) {
        this.syncer.notify(id, false, ms);
      }
    } catch (err) {
      this.log.error({ err }, "presence scan failed");
    }
  }
}

export function setOffline(store: Store, syncer: Syncer, userId: number, now: Date): void {
  const ms = now.getTime();
  store.withWriteTx(() => {
    upsertPresence(store, {
      userId,
      online: false,
      lastSeenAtMs: ms,
      platform: "",
      appVersion: "",
      deviceBrand: "",
      updatedAtMs: ms,
    });
  });
  syncer.notify(userId, false, ms);
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
